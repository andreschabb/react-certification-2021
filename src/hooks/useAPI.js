import { useEffect, useState } from 'react';

import youtube from '../apis/youtube';
import { DEFAULT_SEARCH } from '../utils/constants';

const useAPI = (search = DEFAULT_SEARCH) => {
  const [videos, setVideos] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  useEffect(() => {
    const onSearch = async (searchedText) => {
      try {
        setLoading(true);
        const response = await youtube.get('/search', {
          params: {
            q: searchedText,
          },
        });
        setVideos(response.data.items);
      } catch (e) {
        setError(e);
      } finally {
        setLoading(false);
      }
    };
    onSearch(search);
  }, [search]);

  return [videos, loading, error];
};

export default useAPI;
