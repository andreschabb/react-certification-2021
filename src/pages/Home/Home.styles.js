import styled from 'styled-components';

export const Container = styled.section`
  text-align: center;
  height: 100%;
  width: 80%;
  margin-bottom: 5%;
`;

export const Title = styled.h1`
  font-size: 4rem;
  letter-spacing: -6px;
  font-family: 'Lucida Console', 'Courier New', monospace;
  font-weight: normal;

  @media screen and (max-width: 650px) {
    font-size: 3rem;
    letter-spacing: -4px;
  }

  @media screen and (max-width: 425px) {
    font-size: 2rem;
    letter-spacing: -2px;
  }
`;
