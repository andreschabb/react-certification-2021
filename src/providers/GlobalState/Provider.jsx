import React, { createContext, useContext, useReducer } from 'react';

import reducer from './Reducer';

const initState = {
  isThemeLight: true,
};

const GlobalStateContext = createContext({
  isThemeLight: true,
});

function useGlobalState() {
  const context = useContext(GlobalStateContext);
  if (!context) {
    throw new Error(`Can't use "useTodos" without an TodoProvider!`);
  }
  return context;
}

function GloablStateProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initState);

  return (
    <GlobalStateContext.Provider value={{ state, dispatch }}>
      {children}
    </GlobalStateContext.Provider>
  );
}

export { useGlobalState };

export default GloablStateProvider;
