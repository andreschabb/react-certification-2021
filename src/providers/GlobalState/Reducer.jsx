export default function reducer(state, action) {
  switch (action.type) {
    case 'CHANGE_THEME': {
      return {
        ...state,
        isThemeLight: action.payload,
      };
    }
    default:
      throw new Error('Unkown action');
  }
}
