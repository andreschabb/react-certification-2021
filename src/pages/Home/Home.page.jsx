import React from 'react';

import useAPI from '../../hooks/useAPI';
import VideoList from '../../components/VideoList';
import { Container, Title } from './Home.styles';

const HomePage = () => {
  const [videos, loading] = useAPI();

  return (
    <Container>
      <Title data-testid="home-message">Welcome to the Challenge!</Title>
      {!loading && <VideoList videos={videos} />}
    </Container>
  );
};

export default HomePage;
