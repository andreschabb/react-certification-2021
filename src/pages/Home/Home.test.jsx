import React from 'react';
import { render, screen } from '@testing-library/react';

import useAPI from '../../hooks/useAPI';
import Home from './Home.page';
import { useGlobalState } from '../../providers/GlobalState/Provider';

jest.mock('../../hooks/useAPI');

const mockedVideos = [
  {
    id: { kind: 'youtube#video', videoId: 'nmXMgqjQzls' },
    snippet: {
      title: 'Video Tour | Welcome to Wizeline Guadalajara',
      description:
        'Follow Hector Padilla, Wizeline Director of Engineering, for a lively tour of our office. In 2018, Wizeline opened its stunning new office in Guadalajara, Jalisco, ...',
      thumbnails: {
        medium: {
          url: 'https://i.ytimg.com/vi/nmXMgqjQzls/medium.jpg',
          width: 120,
          height: 90,
        },
      },
    },
  },
  {
    id: { kind: 'youtube#video', videoId: 'nmXMgqjQlhj' },
    snippet: {
      title: 'Title',
      description: 'Description',
      thumbnails: {
        medium: {
          url: 'https://i.ytimg.com/vi/nmXMgqjQzls/medium.jpg',
          width: 120,
          height: 90,
        },
      },
    },
  },
];

jest.mock('../../providers/GlobalState/Provider');

describe('Home', () => {
  beforeEach(() => {
    useGlobalState.mockImplementation(() => ({
      state: { isThemeLight: true },
    }));
  });

  it('should renders without videos', () => {
    useAPI.mockImplementation(() => [[], false]);
    const { container } = render(<Home />);

    expect(screen.getByTestId('home-message')).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('should renders with videos', () => {
    useAPI.mockImplementation(() => [mockedVideos, false]);
    render(<Home />);

    expect(screen.getByTestId('home-message')).toBeTruthy();

    expect(screen.getAllByTestId('video-item').length).toBe(2);
  });
});
