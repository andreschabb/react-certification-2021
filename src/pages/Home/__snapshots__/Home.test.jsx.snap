// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should renders without videos 1`] = `
<div>
  <section
    class="sc-iBPRYJ htirTG"
  >
    <h1
      class="sc-fubCfw kKkKkM"
      data-testid="home-message"
    >
      Welcome to the Challenge!
    </h1>
    <div
      class="sc-gKsewC ikKpsV"
      data-testid="video-list"
    />
  </section>
</div>
`;
